### ISSUE TITLE
Input file format 

### ISSUE THREAD

**Issue by escamoteur** [timfreiheit/r_flutter#3]
Hi,
This looks pretty awesome. Can I only process arb files? 
If would be awesome if we also could process simple cvs files as input for internationalisation. 

Have you thought about this? 
Cheers 
Thomas

**Comment by timfreiheit** [comment]
Hi,
at the moment only `arb` files can be processed. It has started as a simple wrapper around around https://pub.dev/packages/intl_translation but it might not be the best solution. 
Im not sure if the package should support different kind of input formats as this could get quite complex in the end.
`arb` is a json format designed for localization and is suggested by flutter.
There are a lot of different formats to choose from.

Localization is one thing which needs to be improved and I'm open for ideas.

**Comment by szotp** [comment]
I think we should settle on .arb as a standard. It's just a json with structure and everybody undestands json.

This way, for example, it is possible to write VSCode extension that would let you add new keys in place, while writing widgets, without even leaving the dart file. If there was multiple formats it would be much harder.

Things like converting csv to arb, or .xlsx to arb, or google sheet to arb could be done in completely separate package.

**Comment by Mooibroek** [comment]
> This way, for example, it is possible to write VSCode extension that would let you add new keys in place, while writing widgets, without even leaving the dart file. If there was multiple formats it would be much harder.

That would be super awesome. Would really improve the total flutter dev workflow a lot!

**Comment by szotp** [comment]
@Mooibroek I already implemented such feature for this plugin: https://github.com/esskar/vscode-flutter-i18n-json

I may just copy my code, or perhaps they would agree to support r_flutter somehow. Also, their code generator was cool, but was written in typescript so it can't be done as dart package obviously.

**Comment by escamoteur** [comment]
@szotp I think that plugin solves localization pretty good, perhaps there is no need to connect it to r_flutter?

**Comment by timfreiheit** [comment]
the problem with the plugin is that is relies on VSCode. 
When working in a team you should not force the developers to use a specific IDE just for one plugin.
Im thinking of porting the generation logic from VSCode Plugin to r_flutter which might then rely on `arb` instead of `json` files

**Comment by escamoteur** [comment]
Yeah that is a valid point. On the other side a VS code plugin that would let you add entries while you are writing code would be extremely valuable.

**Comment by szotp** [comment]
I created proof of concept for the plugin: https://github.com/szotp/r-flutter-vscode

It will add the key to json file specified in pubspec.yaml and run flutter generate, but flutter analyzer is a bit slow to pick these changes. I already submitted issue to DartCode so they may fix it eventually.

**Comment by szotp** [comment]
Improvements for localization to consider:
1. Add /// documentation with default translation and maybe unprocessed key.
2. Check for missing translations in other translation files. Info could be added to /// doc.
3. Have possiblity to retrieve translation key instead of translated value.

**Comment by escamoteur** [comment]
There is an interesting discussion going on on the Flutter repo maybe it would make sense to join efforts https://github.com/flutter/flutter/issues/41437#issuecomment-537384713

**Comment by timfreiheit** [comment]
The translations should be started soon. I am not very pleased with the current solution.
Maybe this should be done before implementing new features on top of it.

There are two possible ways to implement the localisation from what I see:
* generate code with all translations contained like https://github.com/esskar/vscode-flutter-i18n-json
* generate only accessors which will read the concrete translations from the arb/json files at runtime 
  * this will reduce the generated code and would also make it possible to read the translations from native code if required

What should a solution deliver in my option:
* do not require to add an extra dependency like the `RuntimeArb` or `intl_translation`
* make it possible to change the locale at runtime (this requires the usage of something like `I18n.of(context)` to let flutter know what widgets to rebuild)
* use default locale when the current locale does not provide a certain key
* optional: maybe provide hooks to inject custom translations which could be delivered from remoteconfig to perform a/b testing

**Comment by szotp** [comment]
I would add these to the list of requirements:

* Have possibility to get a translation key (especially important if you want to have dynamic languages with `I18n.of(context)` but have no context instance in your Block, ViewModel, etc.
* Have possibility to translate from a string. For example translating error codes, or keys inside push notification payloads.

**Comment by timfreiheit** [comment]
here is a draft how the generated code could look like:

https://gist.github.com/timfreiheit/b3cb60f08755d6ff1c2e638c7aef785c

**Comment by szotp** [comment]
I'm wondering what is the performance cost of doing a switch like that for every key, compared to overriden methods (like in vscode-flutter-i18n-json).

If it's significant, maybe it would be better to do it their way, and have getString used only when accessing by string. But, customLookup would need to be sacrificed for this to work.

**Comment by timfreiheit** [comment]
I am also not sure about it and I could not find anything about the dart switch internals.
Using methods would make the customLookup harder to implement because it would mean to override every method. But this should not be a huge problem

**Comment by escamoteur** [comment]
Maybe I miss something completely but why not using a Map<language Name, Map<key, translatedString@>>?

**Comment by timfreiheit** [comment]
using a map would require to keep all translations in memory all the time which might be an avoidable waste of resources and would make the usage of hot-reload harder

**Comment by escamoteur** [comment]
You could just load a JSON file from assets and deserialize to a map then you only have one language.
How would switch reduce the needed memory?

**Comment by szotp** [comment]
@escamoteur that's how it works right now

I don't know about memory, but generated methods would have the best performance, and hot reload support also sounds pretty nice. Reloading assets requires at least hot restart.

**Comment by escamoteur** [comment]
Ah, ok in the end we need generated functions to that intellisense can pickup the keys. But the switch inthe example was for my understanding only for a case where the result of a key could still be determined by some other variable

**Comment by timfreiheit** [comment]
there is a first working version at https://github.com/timfreiheit/r_flutter/tree/_master/i18n
check it out if you want.

**Comment by szotp** [comment]
This looks great. My suggestions:

1
Maybe you could use tables in docs, and make missing key more visible (emoji or uppercase letters). For example:
```markdown
  /// |||
  /// ---   |---
  /// __en__|"hello_there"
  /// de    |"hallo_hier"
  /// de_AT |⚠️
  /// pl    |⚠️
  String get hello_there {
    return customLookup?.hello_there ?? _lookup.hello_there;
  }
```

2
It looks like customLookup requires a lot of overhead, even if you don't use it. I probably would handle this in the load method by using different class depending on circumstances: 
if there is no customLookup set, just load concrete `I18nLookup_<code>`. 
If there is customLookup, load some kind of composite lookup that does the same thing with translation as I18n currently does:

```dart
/// customLookup logic extracted from I18n
class CompositeLookup implements I18nLookup {
  final I18nLookup customLookup;
  final I18nLookup _lookup;

  CompositeI18nLookup(this.customLookup, this._lookup);

    String get appName {
           return customLookup?.appName ?? _lookup.appName;
    }
    
    // TODO: rest of keys
  }
}
```
And then there is no need to have I18n and I18nLookup, they could be one class, because custom lookup is handled elsewhere.

**Comment by timfreiheit** [comment]
Hi thank you.
the documentation was improved by using a html table.

I18n and I18nLookup can not be the same class at the moment.
The base I18nLookup delegates to `getString` when it is not implemented. this allows for a generic implementation of a custom lookup without needing to override every method on its own.
The `getString` method of `I18n` delegated to the concrete methods to handle localisation.
Both together would result in an infinite loop.
`return customLookup?.appName ?? _lookup.appName;` seems not to be a performance problem as the `customLookup` is null in this case.

**Comment by szotp** [comment]
I think CompositeLookup would solve this, but I agree that it's not a big issue.

**Comment by timfreiheit** [comment]
I just published the version 0.4.0.
did not saw the comment about VSCode. I will open a new issue for this problem
